Sources/Basics/FileSystem/FileSystem+Extensions.swift

```diff
@@ -211,9 +212,14 @@ extension FileSystem {
 
 extension FileSystem {
     /// SwiftPM directory under user's home directory (~/.swiftpm)
+    /// or under $XDG_CONFIG_HOME/swiftpm if the environmental variable is defined
     public var dotSwiftPM: AbsolutePath {
         get throws {
-            try self.homeDirectory.appending(".swiftpm")
+            if let configurationDirectory = ProcessInfo.processInfo.environment["XDG_CONFIG_HOME"] {
```

Comment:
Let's use [`EnvironmentVariables`](https://github.com/apple/swift-package-manager/blob/main/Sources/Basics/EnvironmentVariables.swift) from this same `Basics` module. With https://github.com/apple/swift-package-manager/pull/7364 `EnvironmentVariables` can also handle case-insensitivity of environment variable keys on Windows. Not sure how applicable XDG is on Windows, but since you're not excluding this logic on that platform, let's handle it properly on all platforms for now.

```suggestion
            if let configurationDirectory = EnvironmentVariables.process()["XDG_CONFIG_HOME"] {
```